Sources/SwiftFusion/Inference/ChordalInitialization.swift

```diff
@@ -0,0 +1,205 @@
+                                          error: Vector9(1.0, 0.0, 0.0, /*  */ 0.0, 1.0, 0.0, /*  */ 0.0, 0.0, 1.0),
+                                          edges: Tuple1(associations[anchorId.perTypeID]!)))
+    
+    return linearGraph
+  }
+  
+  /// This function finds the closest Rot3 to the unconstrained 3x3 matrix with SVD.
+  /// TODO(fan): replace this with a 3x3 specialized SVD instead of this generic SVD (slow)
+  public func normalizeRelaxedRotations(
+    _ relaxedRot3: VariableAssignments,
+    associations: Dictionary<Int, TypedID<Vector9, Int>>,
+    ids: Array<TypedID<Pose3, Int>>) -> VariableAssignments {
+    var validRot3 = VariableAssignments()
+    
+    for v in ids {
+      let M_v: Vector9 = relaxedRot3[associations[v.perTypeID]!]
+      
+      let M = M_v.tensor.reshaped(to: [3, 3])
+      
+      let (_, U, V) = M.transposed().svd(computeUV: true, fullMatrices: true)
+      let UVT: Tensor<Double> = matmul(U!, V!.transposed())
+      
+      let det = (UVT[0, 0].scalar! * (UVT[1, 1].scalar! * UVT[2, 2].scalar! - UVT[2, 1].scalar! * UVT[1, 2].scalar!)
+                 - UVT[1, 0].scalar! * (UVT[0, 1].scalar! * UVT[2, 2].scalar! - UVT[2, 1].scalar! * UVT[0, 2].scalar!)
+                 + UVT[2, 0].scalar! * (UVT[0, 1].scalar! * UVT[1, 2].scalar! - UVT[1, 1].scalar! * UVT[0, 2].scalar!))
+      
+      let R = matmul(matmul(U!, Tensor<Double>(shape: [3], scalars: [1, 1, det]).diagonal()), V!.transposed()).transposed()
+      
+      let initRot = Rot3(coordinate:
+                          Matrix3Coordinate(Matrix3(R[0, 0].scalar!, R[0, 1].scalar!, R[0, 2].scalar!,
+                                                    R[1, 0].scalar!, R[1, 1].scalar!, R[1, 2].scalar!,
+                                                    R[2, 0].scalar!, R[2, 1].scalar!, R[2, 2].scalar!))
+      )
+      
+      // TODO(fan): relies on the assumption of continous and ordered allocation
+      let _ = validRot3.store(initRot)
+    }
+    return validRot3;
+  }
+  
+  /// This function computes the rotations
+  public func computeOrientationsChordal(graph: FactorGraph, val: VariableAssignments, ids: Array<TypedID<Pose3, Int>>) -> VariableAssignments {
+    var relaxedRot3 = VariableAssignments()
+    var associations = Dictionary<Int, TypedID<Vector9, Int>>()
+    
+    for v in ids {
+      let R = val[v].rot.coordinate.R
+      associations[v.perTypeID] = relaxedRot3.store(
+        Vector9(R.s00, R.s01, R.s02,
+        R.s10, R.s11, R.s12,
+        R.s20, R.s21, R.s22))
+    }
+    
+    let R_a = val[anchorId].rot.coordinate.R
+    associations[anchorId.perTypeID] = relaxedRot3.store(Vector9(R_a.s00, R_a.s01, R_a.s02,
+    R_a.s10, R_a.s11, R_a.s12,
+    R_a.s20, R_a.s21, R_a.s22))
+    
+    // regularize measurements and plug everything in a factor graph
+    let relaxedGraph: GaussianFactorGraph = buildLinearOrientationGraph(g: graph, v: val, v_linear: relaxedRot3, associations: associations)
+
+    // Solve the LFG
```

Comment:
Again, I think we should have gone non-linear, and in particular creating factors that just call the Frobenius norm.

Reply:
done